tests/test_mutations.py

```diff
@@ -472,40 +704,226 @@ def test_keep_mutation_parent_zero_rate(self):
+        self.choose_alleles(tables, rng)
+        return tables.tree_sequence()
+
+    def initialise_sites(self, tables):
+        sites = tables.sites
+        mutations = tables.mutations
+        nodes = tables.nodes
+        j = 0
+        for site_id in range(sites.num_rows):
+            site = sites[site_id]
+            mutation_list = []
+            while j < mutations.num_rows and mutations.site[j] == site_id:
+                mut = mutations[j]
+                mutation_item = {'time': nodes.time[mut.node],
+                                 'mutation': mut,
+                                 'new': False}
+                mutation_list.append(mutation_item)
+                j += 1
+
+            mutation_list.reverse()
+            site_item = {'site': site,
+                         'mutation_list': mutation_list}
+            self.sites.append(site_item)
+
+    def populate_tables(self, tables):
+        new_mutations = 0
+        for site_item in self.sites:
+            site = site_item['site']
+            site_id = tables.sites.add_row(**site._asdict())
+            for m in site_item['mutation_list']:
+                mut = m['mutation']
+                parent = mut.parent
+                if parent != tskit.NULL:
+                    parent += new_mutations
+                tables.mutations.add_row(site_id, mut.node, mut.derived_state,
+                                         parent=parent, metadata=mut.metadata)
+                if m['new']:
+                    new_mutations += 1
+
+    def place_mutations(self, tables, rng, discrete_sites=False):
+        # Insert a sentinel into the map for convenience.
+        map_position = np.hstack([self.rate_map.position, [tables.sequence_length]])
+        node_times = tables.nodes.time
+        for edge in tables.edges:
+            branch_start = node_times[edge.child]
+            branch_end = node_times[edge.parent]
+            branch_length = branch_end - branch_start
+            index = np.searchsorted(map_position, edge.left)
+            if map_position[index] > edge.left:
+                index -= 1
+            left = edge.left
+            right = 0
+            while right != edge.right:
+                right = min(edge.right, map_position[index + 1])
+                site_left = np.ceil(left) if discrete_sites else left
+                site_right = np.ceil(right) if discrete_sites else right
+                assert site_left <= site_right
+                assert map_position[index] <= left
+                assert right <= map_position[index + 1]
+                assert right <= edge.right
+                # Generate the mutations.
+                rate = self.rate_map.value[index]
+                mu = rate * (site_right - site_left) * branch_length
+                for _ in range(rng.poisson(mu)):
+                    position = rng.flat(site_left, site_right)
+                    if discrete_sites:
+                        position = np.floor(position)
```

Comment:
Aha! BUT what happens when positions < edge.left? Can easily happen when, say, edge.left = 0.5 and edge.right=0.6. Does your ceiling trickery above rule this out?

Reply:
Perhaps our lives would be easier if we just said you can only generate discrete sites mutations on a tree sequence with discrete edges (and other discrete sites)? Can't really see much to gain from mixing the two, and the semantics are a lot clearer if we don't.